File: 프로그래머스/연습문제/Lv1/공원 산책/공원 산책.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    int x=-1, y=-1;
    for(int i=0; i<park.size(); i++){
        for(int j=0; j<park[0].length(); j++){
            if(park[i][j] == 'S'){
                x = i;
                y = j;
                break;
            }
        }
        
        if(x!=-1) break;
    }
    
    bool flag;
    for(string route : routes){
        char op = route[0];
        int dist = route[2]-'0';
        
        if(op == 'E'){
            if(y+dist >= park[0].size()) continue;
            
            flag = true;
            for(int i=1; i<=dist; i++){
                if(park[x][y+i] == 'X'){
                    flag = false;
                    break;
                }
            }
            
            if(flag) y += dist;
        }
        else if(op == 'W'){
            if(y-dist < 0) continue;
            
            flag = true;
            for(int i=1; i<=dist; i++){
                if(park[x][y-i] == 'X'){
                    flag = false;
                    break;
                }
            }
            
            if(flag) y -= dist;
        }
        else if(op == 'N'){
            if(x-dist < 0) continue;
            
            flag = true;
            for(int i=1; i<=dist; i++){
                if(park[x-i][y] == 'X'){
                    flag = false;
                    break;
                }
            }
            
            if(flag) x -= dist;
        }
        else{
            if(x+dist >= park.size()) continue;
            
            flag = true;
            for(int i=1; i<=dist; i++){
                if(park[x+i][y] == 'X'){
                    flag = false;
                    break;
                }
            }
            
            if(flag) x += dist;
        }
    }
    
    return {x, y};
}

File: 프로그래머스/연습문제/Lv1/공원 산책/공원 산책.py
def solution(park, routes):
    x,y = -1,-1
    for i in range(len(park)):
        for j in range(len(park[0])):
            if park[i][j] == 'S':
                x,y = i,j
                break

        if x != -1: break

    for route in routes:
        op = route[0]
        dist = int(route[2])

        if op == 'E':
            if y+dist >= len(park[0]): continue
            else:
                flag = True
                for i in range(1, dist+1):
                    if park[x][y+i] == 'X':
                        flag = False
                        break

                if flag: y += dist
        elif op == 'W':
            if y-dist < 0: continue
            else:
                flag = True
                for i in range(1, dist+1):
                    if park[x][y-i] == 'X':
                        flag = False
                        break

                if flag: y -= dist
        elif op == 'N':
            if x-dist < 0: continue
            else:
                flag = True
                for i in range(1, dist+1):
                    if park[x-i][y] == 'X':
                        flag = False
                        break
                
                if flag: x -= dist
        else:
            if x+dist >= len(park): continue
            else:
                flag = True
                for i in range(1, dist+1):
                    if park[x+i][y] == 'X':
                        flag = False
                        break

                if flag: x += dist

    return [x, y]
